iree/compiler/Dialect/Flow/Transforms/DeduplicateExecutables.cpp

```diff
@@ -24,111 +25,227 @@ namespace Flow {
 
 namespace {
 
-// Replaces each usage of an entry point with its original symbol name with a
-// new symbol name.
-void replaceEntryPointUses(
-    mlir::ModuleOp moduleOp,
-    const DenseMap<Attribute, SymbolRefAttr> &replacements) {
-  for (auto funcOp : moduleOp.getOps<mlir::FuncOp>()) {
-    funcOp.walk([&](DispatchOp dispatchOp) {
-      auto it = replacements.find(dispatchOp.entry_point());
-      if (it != replacements.end()) {
-        dispatchOp.entry_pointAttr(it->second.cast<SymbolRefAttr>());
-      }
-    });
+template <typename Range, typename Pred>
+bool compare_ranges(Range &&lhs, Range &&rhs, Pred pred) {
+  auto lhsIt = lhs.begin();
+  auto rhsIt = rhs.begin();
+  while (lhsIt != lhs.end() && rhsIt != rhs.end()) {
+    if (!pred(*lhsIt++, *rhsIt++)) return false;
+  }
+  if ((lhsIt == lhs.end()) != (rhsIt == rhs.end())) {
+    // Block count mismatch. We do this here so that we avoid the O(n) scan
+    // that would have been required to calculate the size above.
+    return false;
   }
+  return true;
 }
 
-bool areRegionsEquivalent(Region *lhs, Region *rhs) {
-  if (lhs->getBlocks().size() != rhs->getBlocks().size()) {
-    return false;
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping);
+static bool isStructurallyEquivalentTo(Operation &lhs, Operation &rhs,
+                                       BlockAndValueMapping &parentMapping);
+
+// Recursively compares two regions for structural equivalence.
+// Structural equivalence ensures that operations on both the |lhs| and |rhs|
+// have the same attributes and same use-def structure.
+//
+// Example:
+//   func @lhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//   func @rhs(%arg0 : index) -> index {
+//     %c1 = constant 1 : index
+//     %0 = add %arg0, %c1 : index
+//     return %0 : index
+//   }
+//
+//   assert(isStructurallyEquivalentTo(lhs.getBody(), rhs.getBody()));
+//
+// TODO(#3996): upstream into mlir::OperationEquivalence if this works.
+// TODO(#3996): add symbol ref comparison (add to BlockAndValueMapping).
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs) {
+  BlockAndValueMapping mapping;
+  return isStructurallyEquivalentTo(lhs, rhs, mapping);
+}
+
+static bool isStructurallyEquivalentTo(Region &lhs, Region &rhs,
+                                       BlockAndValueMapping &parentMapping) {
+  // If the region is isolated we don't want to reuse any parent mapping or
+  // pollute it with our mappings.
+  BlockAndValueMapping scopedMapping;
+  BlockAndValueMapping mapping =
+      lhs.isIsolatedFromAbove() ? scopedMapping : parentMapping;
```

Comment:
I don't think that Region::isIsolatedFromAbove is what you want (that one *verifies* whether the region is properly isolated from above (awful name; should probably be "checkIfRegionIsActuallyIsolatedFromAboveAndEmitADiagnosticIfNot")). See my other comments for how to fix this.